### DISCUSSION TITLE
About RIS modeling and configuration

### DISCUSSION THREAD

**Discussion by ekilcioglu** [NVlabs/sionna#661] [Q&A] [answered]
Hello,

My question will be on RIS modeling. As far as I understand from the following text: “For the computation of coverage maps, the ray-based model from [[Vitucci24]](https://nvlabs.github.io/sionna/em_primer.html#vitucci24) is used.”,

the coverage map including the RIS contribution is calculated using the gradient based approach. Does it mean that the coverage map of distance based configuration of the RIS (focusing lens) cannot be calculated or calculated wrongly? If yes, I guess we can somehow create the coverage map of the distance based configured RIS using compute_paths by defining the center of each cell as a receiver, is it correct?

I am asking this because I have an indoor scene to investigate and the coverage map of gradient-based configured RIS seems perfectly fine, I can see the improvement in the path gain while if I configure RIS using focusing lens, the RIS coverage gain becomes almost zero everywhere in the coverage map.

Thank you for your help in advance, best regards.

**Comment by jhoydis** [comment]
Hi @ekilcioglu,

This issue with the focusing lens configuration is that it creates a very narrow peak at the targeted position. This peak is typically much smaller than the size of a cell of a coverage map. For this reason, you simply do not see it.

If the goal of an RIS is to increase coverage in a specific zone, then you should probably not configure it as a focusing lens.

Hope that helps.

**Reply by ekilcioglu** [reply]
Thank you for your response! I understand your reasoning but I already tried to draw my coverage map by using the compute_paths results for both gradient-based and distance-based (focusing lens) approaches and the results for RIS coverage gain are as follows. First of all, I have deployed TX at [3.8,19.7,2.0] m at the top of the scene and RIS at [0.1,6.0,2.0] m on the left side to support two blind zones in terms of the TX-only coverage map, which are the internal room and the right lower side where the transmitter cannot serve enough. The cell sizes are 0.4 m by the way. So, I defined RIS for two modes and I configured it using both approaches. It seems that the distance-based approach results can be seen in the coverage map I have plotted by using the compute_paths results. However, if I use scene.coverage_map() function to plot, the results of distance-based approach doesn't seem while the gradient-based results are totally fine. I am also sending how I plot from the compute_paths results to coverage maps.

So, shouldn't the results be seen as the coverage map derived from compute_paths results or is there something that I am missing?

Coverage map derived from compute paths (Gradient-based configuration)
![Cov_Gain_Compute_path_gradient](https://github.com/user-attachments/assets/1eea05ed-4bd5-48ef-b020-cd33eebb1540)

Coverage map derived from compute paths (Focusing lens configuration)
![Cov_Gain_Compute_path_distance](https://github.com/user-attachments/assets/583c7e66-261f-48e6-8e11-24f7c95eea75)

Coverage map of scene.coverage_map() (Gradient-based configuration)
![Cov_Gain_cov_map_sionna_gradient](https://github.com/user-attachments/assets/48ff8536-bb16-467b-9a5f-2922eadceb3c)

Coverage map of scene.coverage_map() (Focusing lens configuration)
![Cov_Gain_cov_map_sionna_distance](https://github.com/user-attachments/assets/54243498-97dd-4fa3-ac37-d7c73b773858)



Code snippet from compute_paths to coverage map plots:
```py
# Define scene dimensions and resolution
x_dim = 10.0  # meters
y_dim = 20.0  # meters
cell_size = 0.2  # meters

# Compute number of cells in x and y directions
num_cells_x = int(x_dim / cell_size)
num_cells_y = int(y_dim / cell_size)

# Generate receiver coordinates at cell centers with z = 2.0
x_coords = np.arange(cell_size / 2, x_dim, cell_size)
y_coords = np.arange(cell_size / 2, y_dim, cell_size)
coords_RXs = [(x, y, 2.0) for y in y_coords for x in x_coords]

# Add all receivers to the scene
for index, coord_RX in enumerate(coords_RXs):
    self.scene.remove(f"rx_trial-{index+1}")  # Remove previous receiver if necessary
    rx_trial = Receiver(name=f"rx_trial-{index+1}", position=coord_RX)
    self.scene.add(rx_trial)

# Compute paths for all receivers
paths = self.scene.compute_paths(
    max_depth=6, num_samples=int(3e7), los=True,
    reflection=True, diffraction=False, scattering=False, ris=True
)

# Extract and clean path gains
a = tf.squeeze(paths.cir()[0])  # Get channel impulse response
a_real = tf.abs(a)  # Take absolute values
non_nan_mask = tf.math.logical_not(tf.math.is_nan(a_real))  # Mask non-NaN values
a_clean = tf.where(non_nan_mask, a_real, tf.zeros_like(a_real))  # Clean NaNs
path_gain_per_receiver = tf.reduce_sum(a_clean ** 2 * tf.cast(non_nan_mask, a_real.dtype), axis=-1)

# Convert path gain to dB scale
path_gain_per_receiver_dB = 10.0 * tf.math.log(path_gain_per_receiver) / tf.math.log(10.0)

# Map path gains to a 2D matrix
path_gain_matrix = tf.zeros([num_cells_y, num_cells_x], dtype=tf.float32)
for idx, coord_RX in enumerate(coords_RXs):
    cell_x = int(coord_RX[0] // cell_size)
    cell_y = int(coord_RX[1] // cell_size)
    path_gain_matrix = tf.tensor_scatter_nd_update(
        path_gain_matrix,
        indices=[[cell_y, cell_x]],  # Note the order: [y, x]
        updates=[path_gain_per_receiver_dB[idx]]
    )

# Convert the tensor to numpy for visualization
path_gain_matrix_np = path_gain_matrix.numpy()

# Visualization using plt.imshow()
plt.figure(figsize=(8, 16))
plt.imshow(path_gain_matrix_np, origin='lower')
plt.colorbar(label="Path Gain (dB)")
plt.xlabel("X-axis (m)")
plt.ylabel("Y-axis (m)")
plt.title("Coverage Map")
plt.show()
```

Sorry for the long reply and looking forward to your response!

**Comment by ekilcioglu** [comment]
Up

**Comment by jhoydis** [comment]
Hi @ekilcioglu,

I did not have time yet to look into this, but something might cause problems. It seems that the TX, RIS and RX are all located at the same height z=2m. Do you also compute the coverage map at this height, or do you use the default value of z=1.5m?

In principle, you cannot compute a coverage map lying in the x-y plane which has the same z-coordinate as the transmitter.

**Comment by ekilcioglu** [comment]
I didn't change the default value of 1.5m for computing the coverage map provided by Sionna. And yes, all transmitter, RIS center and target points (RXs) are at the same height of 2m in the scene.

Then, maybe the problem with the coverage map of Sionna is to define the target points of the RIS with the same height of the RIS center by myself so that in the focusing lens approach, the power is maximized at the height of 2m for the target points while the coverage map gives the path gain for the height of 1.5m and since the focusing lens approach gives a very narrow peak at the height of 2m and it couldn't reach at the height of 1.5m which is the default value of the coverage map, would that be correct reasoning of this issue? Since there is no problem with the phase gradient approach for the coverage map provided by Sionna, I thought of the reasoning as something like this. If it is correct, then, how did I obtain correct-looking results for the focusing lens approach after defining the coverage-map-cell centers by myself and calculating the path gain for each cell center to create my own coverage map?

**Comment by jhoydis** [comment]
Can you try placing the receivers for computing the coverage via paths at the same height as the coverage map, i.e., 1.5m ?

**Comment by ekilcioglu** [comment]
Yes, I defined the receivers for computing the coverage via paths at the height of 1.5m and the result is below. By the way, I used focusing lens approach to configure the RIS at two target points of still 2 m height. The result is still different from that of computing the coverage map provided by Sionna. However, compared to defining at 2m for the cells, the path gain now looks more like the result of the coverage map provided by Sionna if we define the receivers at the height of 1.5m, but there are still differences compared to "Coverage map of scene.coverage_map() (Focusing lens configuration)", which you can see the result above in my previous post.

Then, should we conclude that if we use focusing lens approach to configure the RIS for the target points at the height of 2m, the RIS coverage gain will disappear at the same position but with the height of 1.5m?

![1-5m](https://github.com/user-attachments/assets/e7c74746-1e4c-4851-8848-bf1d1d777c03)

**Reply by jhoydis** [reply, marked as answer]
Yes. A focusing lens configuration will create constructive interference at precisely one point whereas the phase gradient configuration will simply steer the reflected energy toward a specific direction.

You can also use any other configuration you want. The ones provided with Sionna are just some default configurations.

**Comment by ekilcioglu** [comment]
Yes, I know. Thank you so much for your assistance and your comments Dr. Hoydis!

Best regards
